File: python/day4.py
#!/usr/bin/env python3

with open("../input/day4.txt") as f:
    data = f.read()

required_fields = sorted((
    'byr',
    'iyr',
    'eyr',
    'hgt',
    'hcl',
    'ecl',
    'pid',
))

count = 0
found_fields = {}

def finish_fields():
    global count
    global found_fields

    if 'cid' in found_fields:
        del found_fields['cid']
    if sorted(found_fields.keys()) == required_fields:
        count += 1
    found_fields = {}

for entry in data.splitlines():
    entry = entry.strip()
    if not entry:
        finish_fields()
        continue
    found_fields.update({x.split(':')[0]: True for x in entry.split(' ')})
finish_fields()
print(count)

File: haskell/day4.hs
import qualified Data.List as List
import qualified Data.Map as Map
import qualified Data.Set as Set
import qualified Data.Text as T
import qualified Data.Text.IO as TIO

tuple :: [a] -> (a, a)
tuple (x:y:[]) = (x, y)

parseLine :: [Map.Map T.Text T.Text] -> T.Text -> [Map.Map T.Text T.Text]
parseLine (r:rs) line
    | line == T.empty = Map.empty:r:rs
    | otherwise       = r':rs
    where
        entries =
            map (T.splitOn $ T.singleton ':') $
            T.words line
        r' = Map.union r $ Map.fromList $ map tuple entries

parse :: T.Text -> [Map.Map T.Text T.Text]
parse content =
    reverse $
    List.foldl' parseLine [Map.empty] $
    T.lines $
    content

wantedKeys :: [T.Text]
wantedKeys =
    map T.pack $ List.sort $ [
        "byr",
        "iyr",
        "eyr",
        "hgt",
        "hcl",
        "ecl",
        "pid"
    ]

validPassport :: Map.Map T.Text T.Text -> Bool
validPassport p =
    validKeys
    where
        p' = (Map.delete $ T.pack "cid") $ p
        validKeys = (Map.keys p') == wantedKeys

answer :: [Map.Map T.Text T.Text] -> Int
answer content =
    length $
    filter validPassport $
    content

main :: IO ()
main = do
    content <- TIO.readFile "../input/day4.txt"
    let input = parse content
    putStrLn $ show $ answer input
